p2p/async-buffer/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+/// accepted.
+pub fn new_buffer<T>(capacity: usize) -> (BufferAppender<T>, BufferStream<T>) {
+    let (tx, rx) = unbounded();
+    let sink_waker = Arc::new(AtomicWaker::new());
+    let capacity_atomic = Arc::new(AtomicUsize::new(capacity));
+
+    (
+        BufferAppender {
+            queue: tx,
+            sink_waker: sink_waker.clone(),
+            capacity: capacity_atomic.clone(),
+            max_item_weight: capacity,
+        },
+        BufferStream {
+            queue: rx,
+            sink_waker,
+            capacity: capacity_atomic,
+        },
+    )
+}
+
+/// The stream side of the buffer.
+pub struct BufferStream<T> {
+    /// The internal queue of items.
+    queue: UnboundedReceiver<(T, usize)>,
+    /// The waker for the [`BufferAppender`]
+    sink_waker: Arc<AtomicWaker>,
+    /// The current capacity of the buffer.
+    capacity: Arc<AtomicUsize>,
+}
+
+impl<T> Stream for BufferStream<T> {
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let Some((item, size)) = ready!(self.queue.poll_next_unpin(cx)) else {
+            return Poll::Ready(None);
+        };
+
+        // add the capacity back to the buffer.
+        self.capacity.fetch_add(size, Ordering::AcqRel);
+        // wake the sink.
+        self.sink_waker.wake();
+
+        Poll::Ready(Some(item))
+    }
+}
+
+/// The appender/sink side of the buffer.
+pub struct BufferAppender<T> {
+    /// The internal queue of items.
+    queue: UnboundedSender<(T, usize)>,
+    /// Our waker.
+    sink_waker: Arc<AtomicWaker>,
+    /// The current capacity of the buffer.
+    capacity: Arc<AtomicUsize>,
+    /// The max weight of an item, equal to the total allowed weight of the buffer.
+    max_item_weight: usize,
+}
+
+impl<T> BufferAppender<T> {
+    /// Returns a future that resolves when the channel has enough capacity for
+    /// a single message of `size_needed`.
+    ///
+    /// It should be noted that if there are no items in the buffer then a single item of any capacity is accepted.
+    /// i.e. if the capacity is 5 and there are no items in the buffer then any item even if it's weight is >5 will be
+    /// accepted.
+    pub fn ready(&mut self, size_needed: usize) -> BufferSinkReady<'_, T> {
+        let size_needed = min(self.max_item_weight, size_needed);
+
+        BufferSinkReady {
+            sink: self,
+            size_needed,
+        }
+    }
+
+    /// Attempts to add an item to the buffer returning an error if there is not enough capacity or
+    /// the [`BufferStream`] was dropped.
+    pub fn try_send(&mut self, item: T, size_needed: usize) -> Result<(), BufferError> {
+        let size_needed = min(self.max_item_weight, size_needed);
+
+        if self.capacity.load(Ordering::Acquire) < size_needed {
+            return Err(BufferError::NotEnoughCapacity);
+        }
+
+        let prev_size = self.capacity.fetch_sub(size_needed, Ordering::AcqRel);
```

Comment:
This is the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem) - 2 threads can enter after 118 and wrap the `fetch_sub`.

You need a `compare_exchange` loop for this to be correct.

Reply:
No, this is a single producer, single consumer, so this can't happen

Reply:
Ahhh I missed that, my bad.